### ISSUE TITLE
[Feature] WebAssembly support

### ISSUE THREAD

**Issue by listepo-alterpost** [dart-lang/sdk#32894]
It would be very cool to add support for the [WebAssembly](http://webassembly.org/)

**Comment by terrylucas** [comment]
Yes it would.  This is something we are looking at.  Thank you.

**Comment by Bohne13** [comment]
What, is the effort of webassembly and is there a timeline, for implementation?

**Comment by mraleph** [comment]
There is no timeline and no work is happening on this at the moment.

**Comment by listepo-alterpost** [comment]
@mraleph thanks for answer, but how can I know that already happening?

**Comment by mraleph** [comment]
@listepo-alterpost I am not sure I understand the question. Nothing is happening on this at the moment.

**Comment by b-strauss** [comment]
Using this in flutter apps would be a great usecase.

**Comment by mraleph** [comment]
@b-strauss using it in what way?

**Comment by listepo-alterpost** [comment]
I guess like [Blazor](https://github.com/aspnet/Blazor) alternative(Hummingbird)

**Comment by b-strauss** [comment]
@mraleph 

WebAssembly is a cross platform format. If the DartVM had support for it, we could use the same WASM files on mobile Flutter apps as on web Flutter apps (with Hummingbird via calls to the JS WASM API).

**Comment by Lelelo1** [comment]
Any news on this?

**Comment by mraleph** [comment]
No news.

In general if you are interested in WASM support I recommend including rationale (e.g. "I am interested in WASM because ..."). Knowing use cases that people consider would help us to prioritise things. Thanks.

**Comment by Schonhoffer** [comment]
I am interested in WASM because ..

- I have an existing SPA web app that relies on substantive library that isn't really JavaScript specific, but needs to run client side.
- I want to develop a mobile app with same experience using Flutter, and would have to re-write and maintain a duplicate of this library in Dart alongside the existing JavaScript implementation.
- I would like to be able to re-write that library in a language that targets WASM, and then be able to import that library from both web platform and flutter.

TLDR: code sharing of packages between client-side web & flutter.

**Comment by mraleph** [comment]
@Schonhoffer You can already do this in multiple ways: 

* First option: Dart actually compiles to JavaScript - so if you just rewrite your library in Dart you will be able to use it from both JavaScript and natively from Flutter. [and if you are rewriting it anyway - why not in Dart?]
* Second option: Dart allows you to interoperate with native libraries from Flutter (and this will get even better with FFI that is currently under active development). So you write your program in some native language like C++ (or Rust), compile it natively and use it from Flutter. You can also compile it to WASM and use it on the Web.

As you can see neither of these options requires any sort of WASM support from Dart.

**Comment by nex3** [comment]
Dart Sass is also very interested in WebAssembly. Our biggest user base runs [Dart Sass on Node.js](https://www.npmjs.com/package/sass), but we consistently hear from our users that the performance of the compiled JS output is not up to par. Our users are comparing the performance of our dart2js output to [Node Sass](https://www.npmjs.com/package/node-sass), which is a wrapper around a C++ library. Dart Sass via dart2js is 2x slower than Node Sass in the very best cases, and almost 6x slower when compiling some real-world libraries.

On the other hand, the main reason anyone uses Dart Sass over Node Sass is that it's vastly easier to install a library that doesn't require native compilation. So users have a choice: they can have an easy installation experience or fast compilation, and whichever they choose also comes with a bunch of subtle incompatibilities. It's not a great experience.

WebAssembly would give us the best of both worlds. It would almost certainly represent a major performance boost over dart2js because we don't have to deal with the overhead of JavaScript, and it's just as portable (and thus easy to install) as plain JS. Even if it the WASM binary had to ship with a full garbage collector until WASM supported it, it would make a huge difference to me and my users.

---

As an aside, can we re-open this issue, since the proposal is still under consideration?

**Comment by Bohne13** [comment]
For all of you:
in the following link is a survey about flutter directly from the flutter developers. Because Flutter and Dart are closely linked together there are some questions about Dart and one particular about dart support for web assembly.
Could be interesting?!

https://google.qualtrics.com/jfe/form/SV_3W3aVD2y9CoAe6V?Source=IntelliJ&ClientID=cc844799-def7-4db0-8e50-d20606f2af1d

**Comment by alanjds** [comment]
"Sorry, this survey is not active"

**Comment by alanjds** [comment]
@mraleph The point is about easy interoperability. I can compile Rust/C/C++ to native, but having a to deal with this for every platform is cumbersome and unecessary on early versions of my app.

Do this makes any sense?

**Comment by pie-flavor** [comment]
That and "it already compiles to JavaScript" is not really a big comfort because one of the main perks of using wasm instead of transpiling to JS is the incredible speed increase.

**Comment by mraleph** [comment]
I am reopening the issue to indicate that this is something we are considering, even though we currently have no immediate plans to work on this. 

@pie-flavor 

> one of the main perks of using wasm instead of transpiling to JS is the incredible speed increase.

WASM is not some pixie dust that makes your code magically faster. I actually have some serious doubts that compiling Dart to WASM would bring any speed increase, and in fact expect worse performance due to the lack of dynamic optimizations, insulation from underlying native architecture and inability to rely on builtin optimised primitives and standard library. The only performance characteristic that can potentially be improved by targeting WASM is startup latency. Additionally you will get somewhat predictable performance - but again, I actually doubt it would be better than peak performance you will get from good dart2js code.

**Comment by b-strauss** [comment]
> I am reopening the issue to indicate that this is something we are considering, even though we currently have no immediate plans to work on this.

Are you talking about running WASM binaries on Dart VM, or compiling Dart to WASM?

**Comment by mraleph** [comment]
@b-strauss renamed issue to make it clear.

**Comment by b-strauss** [comment]
@mraleph Is running WASM on the Dart VM also something that is considered? Should there be a separate issue for that?

**Comment by mraleph** [comment]
@b-strauss nothing has changed since my [comment](https://github.com/dart-lang/sdk/issues/32894#issuecomment-455544769) in January. There are no immediate plans for running WASM on Dart VM - because so far there was no strong demand or persuasive use case which would warrant adding such complexity to the VM.

**Comment by alanjds** [comment]
> Is running WASM on the Dart VM also something that is considered? Should there be a separate issue for that?

> There are no immediate plans for running WASM on Dart VM - because so far there was no strong demand or persuasive use case which would warrant adding such complexity to the VM.

What about being able to reuse already-existing business logic coded in a wasm-compilable language?
Compiling to native in 2 or 3 platforms (iOS, Android, Web) is way harder then compile to WASM only and use it on the 3.

To be clear: Your team can have a stance against doing it or even doing it _right now_. But in no way there is "no use case" for running WASM along Dart code, in my opinion.

**Comment by alanjds** [comment]
(cc: @mesquitaa)

**Comment by nex3** [comment]
Although my primary interest here is being able to compile Dart to WASM, I think @alanjds makes a really good point. I expect we're going to be seeing a lot more shared libraries targeting WASM in the next few years, and either Dart's support for that could be a feature that sets it apart from other language options, or its lack of support could be another drawback that makes JS more appealing.

**Comment by pie-flavor** [comment]
@mraleph 
> WASM is not some pixie dust that makes your code magically faster.

Isn't it? That's pretty much the whole point of using a compiled instead of interpreted language. But that's just startup time; the *real* speedup is in not using a language where you can redefine anything anywhere at any moment, because that's the primary thing holding back the JIT. It's not necessarily that the WASM format is special; it's that its end format isn't JS.

**Comment by mraleph** [comment]
@pie-flavor It is without doubt that WASM is a better compilation target than JS for languages like C/C++ and Rust - languages where you manage your memory manually, where your calls are most often statically dispatched and primitive types are unboxed.  

However if you start moving away from this kind of languages towards languages like Dart - which are full of dynamic behaviour, boxed primitive types and GC you will discover it becomes harder to claim with absolute certainty that WASM is a better compilation target than JS - because you would be forced to compile various runtime components (e.g. method dispatch implementation) into WASM  - while in JS you get those things from underlying JS runtime and they are implemented natively and heavily optimised. 

Another interesting thing to consider is that dart2js essentially benefits from two compilation steps - AOT compilation to JS *and* dynamic optimisation of this JS later by JS JIT. If AOT compiler fails to statically specialise some call site, there is still a chance that JS VM would manage to do that. You don't get such luxury with WASM. 

There are a lot of other factors to consider (e.g. builtin libraries - do you want to implement your own array like structure in WASM with associated performance penalty, or do you just want to use heavily optimised native array?), etc, etc.

That said - I don't doubt that there are workloads and programs that would benefit from Dart targeting WASM. All I am saying is that expecting all Dart programs to magically get faster is incorrect.

**Comment by swissonid** [comment]
@pie-flavor here is very good Google I/O talkt about js and WAM. If you don't wanna watch the whole talk. Just jump to 17:38  https://youtu.be/njt-Qzw0mVY

**Comment by armacoty** [comment]
WASM will speed up webgl-based applications.

**Comment by sffc** [comment]
In the Google i18n team, we are considering options to build a single shared library that targets multiple runtimes.  We are evaluating WebAssembly as a compile target, such that our i18n library can be invoked from any VMs that support WASM.  Support for WASM on the Dart VM would be a boon to the impact of our i18n-on-WebAssembly proposal.

**Comment by vsmenon** [comment]
Let's use #37355 for interop with WASM and keep this one for compilation to WASM.

> Dart Sass is also very interested in WebAssembly. Our biggest user base runs Dart Sass on Node.js, but we consistently hear from our users that the performance of the compiled JS output is not up to par. Our users are comparing the performance of our dart2js output to Node Sass, which is a wrapper around a C++ library. Dart Sass via dart2js is 2x slower than Node Sass in the very best cases, and almost 6x slower when compiling some real-world libraries.

@nex3 - a useful data point here (if possible) would be Sass with [Dart AOT](https://dart.dev/tools/dart2aot).  A hypothetical Dart-on-WASM should be slower than that.  I.e., the answer here might be that we need to focus on general Dart AOT performance first.

**Comment by nex3** [comment]
We track Dart Sass benchmarks [here](https://github.com/sass/dart-sass/blob/master/perf.md), including running from a script snapshot, from a native executable, and from JS (our three deploy configurations). AOT performance is competitive with LibSass (written in C++) across the board so we're pretty happy with that. Node.js performance is out big pain point, ranging from about 2x slower than AOT at best to about 5x slower on one real-world benchmark. Even getting it to a *consistent* 2x slower would be a major improvement.

**Comment by vsmenon** [comment]
Thanks, @nex3 .  Just to clarify, by "native executable", you mean via `dart2aot` and `dartaotruntime`?

**Comment by nex3** [comment]
That's right.

**Comment by iapicca** [comment]
As I proposed on Flutter's Github [[1](https://github.com/flutter/flutter/issues/41062)]
if Flutter Web would use WASM instead of dart2js
it would feel "future proof"

**Comment by erlangp** [comment]
+1 WASM

**Comment by alanjds** [comment]
(Hope to not have the thread locked :)

**Comment by mindplay-dk** [comment]
> We track Dart Sass benchmarks [here](https://github.com/sass/dart-sass/blob/master/perf.md), including running from a script snapshot, from a native executable, and from JS (our three deploy configurations).

@nex3 how are you benchmarking these? where's the benchmark script?

**Comment by mraleph** [comment]
@mindplay-dk IIRC `pub run grinder benchmark` runs benchmarking scripts. See https://github.com/sass/dart-sass/blob/master/tool/grind/benchmark.dart for the actual code.

**Comment by insinfo** [comment]
Compiling dart for webassembly would be fantastic

**Comment by johnjames24** [comment]
Compiling dart VM to WASAM would also be interesting

**Comment by Viacheslav-Romanov** [comment]
Are there any updates here guys? 🤔

**Comment by iapicca** [comment]
> How will it exactly speed it up?
> This is entirely wrong. This may be the case in some near future, but not now.
> Right now WASM and JS have the same peak performance.

@nlhnt your assertion is quite debatable,
today as average WASM + native (C / Rust) appears to be 20% faster than JS ([source](https://dev.to/linkuriousdev/to-wasm-or-not-to-wasm-3803)) 
this uplift being worth rewriting a bunch of code base is definitely up to debate, 
but the gain in performance is seems to be a fact.

Moreover comparing a relatively new technology against a super-optimized one
one may thing that there's room for this gap to widen.

I'm not implying that WASM + Dart would reach the performance of C or Rust + WASM,
but a performance uplift has to be expected

**Comment by maks** [comment]
Could the [experimental work that was done for compiling Dart with LLVM](https://medium.com/dartlang/dart-on-llvm-b82e83f99a70) be used as the basis for compiling Dart to WASM?

The kind of use case that this may be useful for is something like the [new Rive wasm runtime](https://twitter.com/rive_app/status/1295452747970486272) which was written in c++, but perhaps, given the other Dart based work Rive is doing, they may have chosen to write it in Dart if there had been a Dart->WASM compiler with a low overhead runtime? 

Given the Dart on LLVM work was based off of Dartino, maybe it would have such a small enough runtime?

**Comment by mraleph** [comment]
> Could the experimental work that was done for compiling Dart with LLVM be used as the basis for compiling Dart to WASM?

Unfortunately not really for a couple of reasons - most importantly because LLVM backend would not be able to target WASM with GC, which is what we are interested in targeting - as opposed to targeting WASM linear memory and compiling our C++ runtime system to WASM. The second reason is that nothing based on Dartino should be considered usable at this point - Dartino comes from Dart 1 days and did not even entirely faithfully implement Dart 1 semantics (e.g. no support for reified generics). 

That being said: given the recent effort (largely driven by V8 team) to implement and evolve GC support in WASM I would say that our interest in WASM as a target for the Dart language is also increased. I can't give any concrete timelines or promises - but I just want to say that WASM is actually starting to look like an interesting target. 

On a tangential note: this was not the only experiment we have done with LLVM - we had intern project implementing LLVM backend for the Dart VM AOT compiler two years ago and most recently an engineer from Alibaba with LLVM background has [implemented one](https://docs.google.com/document/d/15OVvpYKFrvv-nX1CoeO-aXCgjiA38jLh_dGt9Qa8Gnw/edit) as well. The practice shows  that using LLVM does not really have any large benefits.

**Comment by gdelgado11** [comment]
Great!!!, it would be fair and necessary

**Comment by gdelgado11** [comment]
It would be interesting to perhaps redo the C / C ++ part of Dart and Flutter and bring it to Rust. So we would have something with a strong focus on security and performance. And maybe that way we would have Dart compilations in WASM that we would be sure are not going to do strange things or have security holes that make software maintenance so difficult.

**Comment by gdelgado11** [comment]
For the moment and as long as there is no direct Dart compilation to WASM, we can use as an alternative to make modules in Rust which is Multiplatform and through Dart's FFI make calls from Dart / Flutter.

**Comment by maks** [comment]
@mraleph thank you for the very comprehensive reply!

Given your previous comments on JS being a good target based on available built-in GC, it makes perfect sense about WASM.

And thank you for the extra info about the other LLVM experiments, happy to hear that current AOT compileris competitive or 
better than LLVM and I guess given it already supports x64 and arm,I guess adding WASM generation would not be that prohibitive.

My main request for if/when Dart to WASM support is worked on would be: that there be a way of choosing to have a "minimal" runtime. Going by your previous answer to my asking about the VM runtime, I hope that excluding dart:io support, icu and WASM having builtin GC, it could mean that the Dart VM runtime could be quite small for a WASM target.

**Comment by aminya** [comment]
Dart is already compiled to JavaScript. So why not compile it to AssemblyScript instead! AssemblyScript has already implemented GC, WASM interop (using as-binder), and it has solved many of the issues.

You just need to add a few more things to the current Dart to JavaScript compiler.

https://www.assemblyscript.org/

**Comment by mraleph** [comment]
> You just need to add a few more things to the current Dart to JavaScript compiler like passing the types.

If you try to add these "few more things" yourself you will discover that JavaScript and AssemblyScript are two very different targets despite some syntactic similarities that latter has to TypeScript.

> this project already implements a Dart to TypeScript.

This project implements subset of Dart to TypeScript, and a single pass AST conversion is not something you want to be doing really doing - the devil is in the details. That's why dart2js is hundred of thousands of lines of code.

**Comment by mraleph** [comment]
> The end users will certainly appreciate "a little more" progress on this even if it is based on a compiler that supports a subset of Dart. 

Unfortunately we are not interested in a compiler that supports subset of Dart. We are interested in a compiler that can run actual Dart applications (which are usually thousands of lines of Dart code using most of its language features).

That however should not discourage you from building a compiler that supports subset and satisfies your particular use case. e.g. similar to how somebody built that dart2ts compiler you are linking to - which is far from something that we would ever consider doing ourselves.

**Comment by aminya** [comment]
You can't compile all the Dart applications to WebAssembly because WASM currently does not have WebIDL bindings (or it will have [a low performance like how Blazer tries to do so](https://krausest.github.io/js-framework-benchmark/current.html)).

To **fully** compile all Dart applications, you need to wait for this issue to be resolved and then appear in the browsers. That seems like "a few years" from now. 
https://github.com/WebAssembly/proposals/issues/8

**Comment by StEvUgnIn** [comment]
The problem is that Wasm isn't targetable as it does not integrate a solution for garbage collection... Thus which limits Webassembly to low level languages being C / C++ / C# / Go / Rust and some others. 

With Wasm 2.0 next generation languages like Python / Dart / D / Java could be used.

**Comment by maks** [comment]
> The problem is that Wasm isn't targetable as it does not integrate a solution for garbage collection

I don't think that's actually correct. Rather it means the full DartVM along with it's GC needs to be included in the WASM which of course significantly increases the size.

**Comment by aminya** [comment]
> Rather it means the full DartVM along with it's GC needs to be included in the WASM which of course significantly increases the size.

That's true. GC is just an algorithm that can be implemented in the wasm binary. The GC will take some space, but for example, AssemblyScript does that and it has one the smallest generated wasm binaries among the others.

**Comment by StEvUgnIn** [comment]
> That's true. GC is just an algorithm that can be implemented in the wasm binary. The GC will take some space, but for example, AssemblyScript does that and it has one the smallest generated wasm binaries among the others.
> 
> [#32894 (comment)](https://github.com/dart-lang/sdk/issues/32894#issuecomment-684808186)

You may start a Dart-to-AssemblyScript compiler project if you want, but remember that AssemblyScript is not a standard as Dart, WebAssembly, JavaScript, TypeScript and EcmaScript are!

I am calling the official AssemblyScript documentation here:
> Also, both can make use of compiler infrastructure that has been created by many individuals and corporations over years. In contrast, AssemblyScript is a relatively young project with limited resources that strives to create a viable alternative from another perspective.

**Comment by iapicca** [comment]
@StEvUgnIn 

> I am calling the official AssemblyScript documentation here:

Since the scope of your request is clearly different
I think your feature request should be in a separate issue, 
rather than trying to hijack this one

**Comment by xxgreg** [comment]
Any public comments about the work going on here?

https://github.com/dart-lang/sdk/commits/wasm-backend

**Comment by mraleph** [comment]
> Any public comments about the work going on here?

Long story short: over the last 3 months we have hosted an intern who worked on targeting WASM+GC. He got some preliminary results: managed to implement enough of a backend working to translate a simple merge-sort benchmark. Unfortunately internship time has just run out. We are interested in driving this further, but right now we have no concrete plans, things are in discussion.

/cc @mkustermann

**Comment by yarcowang** [comment]
Glad to see Dart can be compiled into WASM~ +1
The browser as the desktop environment seems will happen very soon.
Dart to  WASM (but not rely on another scripting language: javascript)  seems more wisdom because javascript is not designed for big programs(at least that's why they now have typescript).

**Comment by mymikemiller** [comment]
Another use case: I'd like to compile Dart to Wasm so I can write canisters that run on [DFINITY's Internet Computer](https://dfinity.org/). Only [languages that compiles to Wasm](https://sdk.dfinity.org/docs/developers-guide/work-with-languages.html) are viable, so Dart can not be used to create canisters for the Internet Computer until this support is added.

**Comment by StEvUgnIn** [comment]
> Another use case: I'd like to compile Dart to Wasm so I can write canisters that run on [DFINITY's Internet Computer](https://dfinity.org/). Only [languages that compiles to Wasm](https://sdk.dfinity.org/docs/developers-guide/work-with-languages.html) are viable, so Dart can not be used to create canisters for the Internet Computer until this support is added.

Dart is a very good standard but I invite you to use TypeScript. The syntax is very closed and you might as well find similarities (Futures, class, typing, access level, etc.).

**Comment by 00ff0000red** [comment]
> You may start a Dart-to-AssemblyScript compiler project if you want, but remember that AssemblyScript is not a standard as Dart, WebAssembly, JavaScript, TypeScript and EcmaScript are!

Since when were JavaScript and EcmaScript different standards?
And when did *TypeScript* get a standard? Dart is approved by the ECMA standard under [ECMA-408](http://www.ecma-international.org/publications/standards/Ecma-408.htm), unlike TS.

**Comment by 00ff0000red** [comment]
We have dart2native, which supposably can compile Dart code to natively executable code, without dependencies, although, it does add some code for type checking and GC, why can't the same be done for Wasm? A large, inefficient Wasm binary is a start, it's not a great start, but a start nonetheless.

**Comment by mhoad** [comment]
I was looking at a use case for using wasm as a module on the server. CNCFs Open Policy Agent project ships in three varieties 
1. As a go library
2. As a stand-alone REST server
3. As a WASM module.

Right now my only option is the number two which seems to introduce a lot of additional overhead. I think WASM could make sense as a target for a lot of server side application components and the ability to handle them natively in Dart would be a huge improvement for developer experience and interoperability.

**Comment by cedvdb** [comment]
There is a package that uses c ffi on pub dev that I'd like to use. Unfortunately that means I wouldn't be able to use it on the web. 

Would wasm support mean I could in fact use it on the web by compiling that C into wasm ? This is a bit unclear to me.

**Comment by mraleph** [comment]
@cedvdb 

> Would wasm support mean I could in fact use it on the web by compiling that C into wasm ? This is a bit unclear to me.

No, this is an orthogonal problem. What you want is `dart:ffi` implementation on the Web using WASM.

Note that if you have a library that compiles to WASM you can already use it on the Web - it's just that you need to write bindings twice (once using `dart:ffi` and once using something like `dart:js`/`package:js`). 

We have discussed possibility of providing Web version of `dart:ffi` - but so far we did not have resources to do that.

**Comment by levifeldman** [comment]
The Motoko Language, built for webassembly is becoming open source next week, it has built in garbage collection and compiles to wasm maybe dart can use the garbage collecter from the Motoko project for the wasm.?

https://sdk.dfinity.org/docs/language-guide/motoko.html
https://stackoverflow.blog/2020/08/24/motoko-the-language-that-turns-the-web-into-a-computer/

**Comment by CrimsonCodes0** [comment]
> The Motoko Language, built for webassembly is becoming open source next week, it has built in garbage collection and compiles to wasm maybe dart can use the garbage collecter from the Motoko project for the wasm.?
> 
> https://sdk.dfinity.org/docs/language-guide/motoko.html
> https://stackoverflow.blog/2020/08/24/motoko-the-language-that-turns-the-web-into-a-computer/

I didn't develop Dart, but I'm still fairly certain that the GCs would be incompatible.

**Comment by mraleph** [comment]
As explained before: compiling Dart to linear memory Wasm and implementing GC on top of linear memory is not really a problem - if we had infinite resources and nothing else to do then we would most certainly have already done it. However with the current set of resource constraints and priorities we are not really interested in targeting it - because it is not going to solve any problems that we are facing.  

Nevertheless, we do see Wasm GC as a really interesting target for Dart-for-Web deployments, with a potential to achieve better startup / efficiency than what is possible with `dart2js`.  We are currently engaging with Wasm GC effort and are building a compilation toolchain capable of compiling some meaningful benchmarks. Once we have some more concrete results we will share them.

**Comment by wanton7** [comment]
Waiting for Wasm GC makes total of sense. But Wasm target is not just meaningful to browsers. There are also interesting upcoming server-side Wasm runtimes like https://github.com/lunatic-solutions/lunatic . It tries to build Erlang type of cheap isolated processes and reliability top of Wasm. Then there is https://github.com/faasm/faasm a high-performance stateful serverless runtime. I bet in future we see Wasm targeted lot more in server side solutions where publish your server-side app as Wasm instead of running you app in containers or as raw executable.

**Comment by CrimsonCodes0** [comment]
@wanton7 Do note that Wasm GC is a Wasm feature, not a browser feature, server-side Wasm engines are allowed to support the GC extension too. This means that Dart compiled to Wasm could still be executed outside of a browser.

**Comment by wanton7** [comment]
@CrimsonCodes0 Yes I know ;) I thought from your comment `Dart-for-Web deployments` meant just browsers. I just wanted to mention Wasm in future won't be browser only target.

**Comment by levifeldman** [comment]
> Nevertheless, we do see Wasm GC as a really interesting target for Dart-for-Web deployments, with a potential to achieve better startup / efficiency than what is possible with `dart2js`. 

for the more, dart->wasm compile would get rid of the platform(dartvm,dart2js,..)-differences like the flutter-web int: 2^53 and dartvm int: 2^64 and the incompatibility of the BigInt cross platforms... .

**Comment by vsmenon** [comment]
We're certainly excited about the potential here and engaged with the WASM GC effort.  But, just to be clear, there is a long road ahead still:  https://github.com/WebAssembly/gc/blob/master/proposals/gc/Overview.md

**Comment by unicomp21** [comment]
@vsmenon does Dart really need WASM GC?  Why must we wait on WASM GC?

Would it be possible for a dart app to declare how much memory it needs at build time?  And from here memory gets preallocated/managed in the app?  Using this memory pool, if needed, other virtual memory could be simulated, via a remote storage interface and some sort of paging?

**Comment by mraleph** [comment]
@unicomp21 see https://github.com/dart-lang/sdk/issues/32894#issuecomment-831774883 - Dart could in fact target WASM MVP (aka WASM with linear memory), but we are not interested in pursuing such translation at the moment.

**Comment by unicomp21** [comment]
@mraleph are the "problems" enumerated somewhere?

**Comment by mraleph** [comment]
> @mraleph are the "problems" enumerated somewhere?

Small download size and high efficiency.

**Comment by gmpassos** [comment]
Waiting for an alpha version and SIMD support! 😎

**Comment by insinfo** [comment]
@gmpassos
I'm looking forward to that too, and I think to use SIMD efficiently in dart would be a language construct like "unsafe context" that I've proposed here https://github.com/dart-lang/language/issues/1758#issuecomment-887669522

**Comment by iapicca** [comment]
awesome [article](https://medium.com/dartlang/experimenting-with-dart-and-wasm-ef7f1c065577) @mit-mit

**Comment by insinfo** [comment]
I have tested this dart2wasm prototype and it is very promising.
https://github.com/askeksa-google/sdk/blob/wasm_prototype/dart2wasm.md

D:\MyDartProjects\dart-sdk\sdk>v8\d8.exe fibonacci.js
0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987
Execution time: 4.5390000000000015 ms

D:\MyDartProjects\dart-sdk\sdk>dart fibonacci.dart
0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987
Execution time: 0:00:00.002000

D:\MyDartProjects\dart-sdk\sdk>v8\d8.exe --experimental-wasm-gc pkg/dart2wasm/bin/run_wasm.js -- fibonacci.wasm
0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987
Execution time: 0:00:00.003000

**Comment by whoizit** [comment]
> Execution time: 0:00:00.002000

is it 2ms?

**Comment by ykmnkmi** [comment]
Interaction with the browser and the DOM?

**Comment by rohansingh** [comment]
Definitely see the issues with lack of WASM GC, but figured I'd add another use case here that has nothing to do with web.

At @tidbyt have a connected hardware product with a companion app. The companion app for the hardware is written in Flutter so the team is familiar with Dart. The firmware for the hardware is written in C/C++, which runs on a microcontroller.

Ideally we'd like to write parts of the firmware in a higher-level language than C/C++. We'd also want these parts to be easily portable to other microcontrollers and architectures. As a proof-of-concept, I've added a WASM runtime to our firmware and compiled experiments in a couple of different languages to WASM. These languages have different levels of maturity and learning curves, and I'm still evaluating what would work best.

If Dart compiled to WASM, we could use a language our team already knows and share code for some features between our Flutter app and our embedded device.

Of course this would also be contingent on having reasonably-sized WASM binaries that fit in the embedded environment. It's possible we could achieve something similar by embedding a JS interpreter and using `dart2js`, but that might also run into resource constraints. Either way we probably only have space for either a WASM runtime or a JS runtime on an embedded device.

**Comment by maks** [comment]
I have a similar use case to @rohansingh in wanting to use WASM on small devices and microcontrollers and at the moment I have to use eg. Assemblyscript, but would much rather use Dart. Yes I realise Dartino was abandoned, but I really do think there is a nice usecase for using Dart on devices as @rohansingh mentioned for teams and developers  that are already doing companion apps for the devices in Flutter.

Given this usecase, I would be writing Dart code in a similar style to that in AS and taking care of memory management myself, so I tried to find if there was a existing way in Dart to aot compile with GC disabled. I haven't been able to see anything to do this or perhaps I am missing something? 

I guess what I'd like to do is what the dart2wasm prototype does but with GC removed -  @mraleph is that likely to be possible or am I missing something fundamental here?

**Comment by mraleph** [comment]
> I guess what I'd like to do is what the dart2wasm prototype does but with GC removed - @mraleph is that likely to be possible or am I missing something fundamental here?

I don't think that would ever be possible, Dart is fundamentally a GC managed language. There are allocations hidden from the user in the core library and there are allocations that you don't even think about as a programmer (e.g. you evaluate a sum of two integers and put the result into a `List` - that can trigger an allocation of a box).

You could build a dialect of Dart which eliminates hidden allocations and allows you to manage memory manually - but that's not something we ever plan to do ourselves.

**Comment by ElijahLynn** [comment]
My use case for WASM support in Dart is to use this library here, https://github.com/mfcc64/showcqt-js/blob/master/showcqt.js,  which is an FFMpeg Audio Spectrum Visualizer that has been converted to WASM. Would be great to use the WASM port so I can do the web app too.

**Comment by mit-mit** [comment]
@ElijahLynn please do try it out, and let us know how it goes!

**Comment by loeffel-io** [comment]
+1

**Comment by ykmnkmi** [comment]
@liudonghua123 
`Dart SDK version: 2.19.0-374.1.beta (beta)`
`dartaotruntime  path-to-sdk/bin/snapshots/dart2wasm_product.snapshot -h`

**Comment by askeksa-google** [comment]
> I always got `Error when reading 'pkg/dart2wasm/bin/dart2wasm.dart': No such file or directory.` errors.

You need to run the command in a checkout of the [SDK repository](https://github.com/dart-lang/sdk).

Added a clarification to the text: https://dart-review.googlesource.com/c/sdk/+/271881

**Comment by ykmnkmi** [comment]
@askeksa-google I'm trying to run WASM module in the browser (dart), but I got error like this `CompileError: wasm validation error: at offset 13`.

**Comment by askeksa-google** [comment]
> @askeksa-google I'm trying to run WASM module in the browser, but I got error like this `CompileError: wasm validation error: at offset 13`.

The Wasm extensions we depend on for running Wasm code produced by dart2wasm are still experimental and in flux. The [experimental instructions](https://github.com/dart-lang/sdk/blob/main/pkg/dart2wasm/dart2wasm.md) indicate how such a Wasm file can be run using `d8`, but this way of running it will also change and will sometimes only work on specific version ranges of `d8`.

**Comment by peterscodee** [comment]
Updates: https://twitter.com/kevmoo/status/1600973667638607872

**Comment by Keithcat1** [comment]
Will Dart support WASI? It'd be able to run anywhere there's a WASM runtime like Wasmer or Wasmtime, maybe one day users can just download one .wasm file and not have to worry about the getting the wrong platform / architecture.

**Comment by jodinathan** [comment]
@mraleph 

> You could build a dialect of Dart which eliminates hidden allocations and allows you to manage memory manually

How so?  
For clarification, do you have any examples please?

**Comment by ykmnkmi** [comment]
@jodinathan look at Vala lang.

**Comment by StEvUgnIn** [comment]
This developer @NachoBrito actually shows that Dart needs to focus on Webassembly instead of JavaScript. JS Interop needs work  to support the new HTML5 APIs:
https://github.com/NachoBrito/flutter-multiapp-poc/blob/main/component1/web/js/interop.js

**Comment by unicomp21** [comment]
Will dart webassembly support using the upcoming tail call support?  My reason for asking.  Looking at the "out of order execution" supported by modern cpu's, will tail call's essentially allow us to await cache line misses?  And run something else in the meantime where the cache lines have been loaded?

**Comment by kevmoo** [comment]
> Will dart webassembly support using the upcoming tail call support? My reason for asking. Looking at the "out of order execution" supported by modern cpu's, will tail call's essentially allow us to await cache line misses? And run something else in the meantime where the cache lines have been loaded?

@askeksa-google ?

**Comment by askeksa-google** [comment]
> Will dart webassembly support using the upcoming tail call support? My reason for asking. Looking at the "out of order execution" supported by modern cpu's, will tail call's essentially allow us to await cache line misses? And run something else in the meantime where the cache lines have been loaded?

The tail call instructions are just fused call-and-return instructions that pop the caller's stack frame before pushing the callee's stack frame. While this is a useful primitive in some coroutine implementations, the tail call instructions do not by themselves confer any capability to suspend and resume the execution of a function body. Such capabilities are being discussed as part of the [Stack Switching Proposal](https://github.com/WebAssembly/stack-switching).

Dart has the `sync*` feature for writing generator functions, which have the capability to be suspended and resumed somewhat similarly to what you suggest. The current implementation of these in the Wasm backend (and in the other backends as well) has some significant overhead associated with it, though, mainly from having to keep the local state of the function alive in heap objects across suspension points. A future Wasm-level stack switching feature could potentially reduce some of that overhead.

**Comment by unicomp21** [comment]
@askeksa-google I'm confused.  Why are the leaningtech guys realizing large performance boosts by having tail calls?

https://leaningtech.com/extreme-webassembly-2-the-sad-state-of-webassembly-tail-calls/
https://leaningtech.com/fantastic-tail-calls-and-how-to-implement-them/

**Comment by askeksa-google** [comment]
Maybe I misunderstood what you wanted to do with the tail calls. It's certainly feasible to add an option to emit tail calls whenever the result of a call is directly returned. This will make stack traces different from what would be expected from the Dart source, but this is already the case for other optimizations, such as inlining.

It may indeed improve performance in some cases. I don't see what it has to do with hiding cache misses, though.

**Comment by unicomp21** [comment]
@askeksa-google  In talking w/ some of the redpanda engineers, my understanding is c++ co_await essentially makes cache line misses "await'able".  Another strand/coroutine can be resumed where the cache line has been loaded.

Per Gor, tail calls are required in order to support co_await in webassembly.  I'm wondering if dart await can perform as efficiently as c++ co_await, in webassembly?

https://www.youtube.com/watch?v=_fu0gx-xseY&t=11s

**Comment by kevmoo** [comment]
@liudonghua123 we're not ready to release things beyond preview. 

Keep in mind, all of the runtime support for WasmGC is also in preview.

flutter.dev/wasm has the latest details

**Comment by liudonghua123** [comment]
> @liudonghua123 we're not ready to release things beyond preview.
> 
> Keep in mind, all of the runtime support for WasmGC is also in preview.
> 
> https://flutter.dev/wasm has the latest details

Thanks, I also found some docs on https://github.com/dart-lang/sdk/tree/main/pkg/dart2wasm.

**Comment by iapicca** [comment]
@kevmoo 
will the "final" version of the feature require d8 as it does now?

**Comment by kevmoo** [comment]
@iapicca – "final" is tricky to define. We'll certainly need a runtime that supports WasmGC. It's possible we may make our first release require a JS runtime as well, since we're VERY focused on the Flutter+Browser scenario. Our support will certainly evolve, though. I'd love to think we'd run on most Wasm runtimes (as long as they support GC), but that's not our focus now or in the near future.

**Comment by nex3** [comment]
I'm also _extremely_ interested in a more portable WASM output. Requiring GC is totally understandable, but it's very important for our use-cases that we be able to integrate with as many WASM runtimes as possible—even if it means using a much lower-level API to communicate out.

**Comment by iapicca** [comment]
@kevmoo I was attempting to make a dart package to use [fermyon spin](https://www.fermyon.com/wasm-languages/dart),
but this is kinda of a blocker
I understand a true dart2wasm compilation is blocked by [wasmGC](https://github.com/WebAssembly/gc),
I just hope that once is out we don't keep depending on flutter related approaches
kneecapping any possible "pure dart" implementation

**Comment by mit-mit** [comment]
We've now landed new support in the `dart` cli for invoking the dart 2 wasm compiler in the Dart and Flutter main channels:

```
$ dart help compile wasm
Compile Dart to a WebAssembly/WasmGC module (EXPERIMENTAL).

Usage: dart compile wasm [arguments] <dart entry point>
-h, --help              Print this usage information.
-o, --output            Write the output to <file name>.
                        This can be an absolute or relative path.
    --[no-]optimize     Optimize wasm output using Binaryen wasm-opt.
                        (defaults to on)
-v, --verbose           Print debug output during compilation
    --enable-asserts    Enable assert statements.
```

I have a small example of how this can be used for a Dart web app here:
https://github.com/mit-mit/sandbox/tree/main/demos/webwasm

Note that this is still within our initial focus of using WasmGC in the context of web apps (whether Flutter web apps or general ones). Over time we'd like to support more general WasmGC execution -- this issue tracks some of the additional work required for that: https://github.com/dart-lang/sdk/issues/53884

**Comment by liudonghua123** [comment]
It seems I can compile dart to wasm with dart 3.4.0.

```
Liu.D.H  dart2wasm-hello   7ms  15:37 > cat > hello.dart
void main() {
  print("Hello, World!");
}

Liu.D.H  dart2wasm-hello   17.767s  15:38 > dart hello.dart
Hello, World!

Liu.D.H  dart2wasm-hello   3.995s  15:38 > dart --version
Dart SDK version: 3.4.0 (stable) (Mon May 6 07:59:58 2024 -0700) on "windows_x64"

Liu.D.H  dart2wasm-hello   1.434s  15:38 >
Liu.D.H  dart2wasm-hello   337ms  15:39 > dart help compile wasm
Compile Dart to a WebAssembly/WasmGC module (EXPERIMENTAL).

Usage: dart compile wasm [arguments] <dart entry point>
-h, --help                  Print this usage information.
-o, --output                Write the output to <file name>.
                            This can be an absolute or relative path.
-v, --verbose               Print debug output during compilation
    --enable-asserts        Enable assert statements.
-D, --define=<key=value>    Define an environment declaration. To specify multiple declarations, use multiple options or
                            use commas to separate key-value pairs.
                            For example: dart compile wasm -Da=1,b=2 main.dart

Run "dart help" to see global options.

Liu.D.H  dart2wasm-hello   2.156s  15:40 > dart compile wasm hello.dart -o hello.wasm
*NOTE*: Compilation to WasmGC is experimental.
The support may change, or be removed, with no advance notice.

Generated wasm module 'hello.wasm', and JS init file 'hello.mjs'.

Liu.D.H  dart2wasm-hello   9.375s  15:41 >
```

But I can't run it with wasmer or wasmedge.

```
Liu.D.H  dart2wasm-hello   16ms  15:43 > node hello.mjs

Liu.D.H  dart2wasm-hello   1.027s  15:43 > wasmer hello.unopt.wasm
error: Unable to compile "hello.unopt.wasm"
╰─▶ 1: compile error

Liu.D.H  dart2wasm-hello   39ms  15:44 > wasmedge hello.wasm
[2024-05-21 15:44:21.194] [error] loading failed: integer representation too long, Code: 0x36
[2024-05-21 15:44:21.197] [error]     Bytecode offset: 0x0000000c
[2024-05-21 15:44:21.198] [error]     At AST node: function type
[2024-05-21 15:44:21.199] [error]     At AST node: type section
[2024-05-21 15:44:21.199] [error]     At AST node: module
[2024-05-21 15:44:21.200] [error]     File name: "D:\\code\\dart\\dart2wasm-hello\\hello.wasm"

Liu.D.H  dart2wasm-hello   385ms  15:44 > wasmer hello.wasm
error: Unable to compile "hello.wasm"
╰─▶ 1: compile error

Liu.D.H  dart2wasm-hello   36ms  15:44 >
```

**Comment by iapicca** [comment]
> It seems I can compile dart to wasm with dart 3.4.0.
> 
> [...]
> 
> But I can't run it with wasmer or wasmedge.

@liudonghua123 
because compile only to wasm js runtime see 
- https://github.com/dart-lang/sdk/issues/53884

**Comment by yanglaolee** [comment]
Wen dart to wasm runtime??

--2024.05.23

**Comment by mit-mit** [comment]
This feature is now considered done. There are some planned improvements outside the core though. For current status, please see https://dart.dev/web/wasm

**Comment by mkustermann** [comment]
For those looking for future things
* Interacting with other wasm modules: [Issue 55856](https://github.com/dart-lang/sdk/issues/55856 )
* Interacting with C code compiled to wasm: [Issue 46690](https://github.com/dart-lang/sdk/issues/46690)
* Running dart2wasm-compiled code in non-JS environments: [Issue 53884](https://github.com/dart-lang/sdk/issues/53884)
